src/lib/syncable_ledger/syncable_ledger.ml

```diff
@@ -427,58 +515,83 @@ end = struct
-      children for retrieval if the values in the underlying ledger don't match
-      the hashes we got from the network. *)
-  let add_child_hashes_to :
+  (* Merges each 2 contigous nodes, halving the size of the array *)
+  let merge_siblings : Hash.t array -> index -> Hash.t array =
+   fun nodes height ->
+    let len = Array.length nodes in
+    if len mod 2 <> 0 then failwith "length must be even" ;
+    let half_len = len / 2 in
+    let f i = Hash.merge ~height nodes.(2 * i) nodes.((2 * i) + 1) in
+    Array.init half_len ~f
+
+  (* Assumes nodes to be a power of 2 and merges them into their common root *)
+  let rec merge_many : Hash.t array -> index -> Hash.t =
+   fun nodes height ->
+    let len = Array.length nodes in
+    match len with
+    | 1 ->
+        nodes.(0)
+    | _ ->
+        let half = merge_siblings nodes height in
+        merge_many half (height + 1)
+
+  let merge_many : Hash.t array -> index -> index -> Hash.t =
+   fun nodes height subtree_depth ->
+    let bottom_height = height - subtree_depth in
+    let hash = merge_many nodes bottom_height in
+    hash
+
+  (* Adds the subtree given as the 2^k subtree leaves with the given prefix address *)
+  (* Returns next nodes to be checked *)
+  let add_subtree :
          'a t
       -> Addr.t
-      -> Hash.t
-      -> Hash.t
-      -> [ `Good of (Addr.t * Hash.t) list
-           (** The addresses and expected hashes of the now-retrievable children *)
+      -> Hash.t array
+      -> int
+      -> [ `Good of (Addr.t * Hash.t) array
          | `Hash_mismatch of Hash.t * Hash.t
-           (** Hash check failed, peer lied. First parameter expected, second parameter actual. *)
-         ] =
-   fun t parent_addr lh rh ->
-    let ledger_depth = MT.depth t.tree in
-    let la, ra =
-      Option.value_exn ~message:"Tried to fetch a leaf as if it was a node"
-        ( Or_error.ok
-        @@ Or_error.both
-             (Addr.child ~ledger_depth parent_addr Direction.Left)
-             (Addr.child ~ledger_depth parent_addr Direction.Right) )
-    in
-    let expected =
-      Option.value_exn ~message:"Forgot to wait for a node"
-        (Addr.Table.find t.waiting_parents parent_addr)
-    in
-    let merged_hash =
-      (* Height here is the height of the things we're merging, so one less than
-         the parent height. *)
-      Hash.merge ~height:(ledger_depth - Addr.depth parent_addr - 1) lh rh
-    in
-    if Hash.equal merged_hash expected then (
-      (* Fetch the children of a node if the hash in the underlying ledger
-         doesn't match what we got. *)
-      let should_fetch_children addr hash =
-        not @@ Hash.equal (MT.get_inner_hash_at_addr_exn t.tree addr) hash
+         | `Invalid_length ] =
+   fun t addr nodes requested_depth ->
+    let open (val t.context) in
+    let len = Array.length nodes in
+    let is_power = Int.is_pow2 len in
+    let is_more_than_two = len >= 2 in
```

Comment:
Nit: `is_more_than_two` is misleading, `2` is not more than 2, but gives `is_more_than_two = true`.